mitodl/mit-open
# Change primary key field for department and offeror

# What are the relevant tickets?
#258 

# Description (What does it do?)
Via migrations, changes primary keys:

`LearningResourceDepartment`: from `id` to `department_id`
`LearningResourceOfferor`: from `id` to `code`

and updates foreign keys on `LearningResource` so they point to the correct values.

# How can this be tested?
- On the main branch, make sure you have departments, offerors, and platforms populated as well as some learning resources:
  ```
  ./manage.py update_platforms && ./manage.py update_departments && ./manage.py update_offered_by
  ./manage.py backpopulate_ocw_data --course-name 6-001-structure-and-interpretation-of-computer-programs-spring-2005
  ./manage.py backpopulate_prolearn_data
  ```
- Switch to this branch and run the migrations (either `docker compose up` or `./manage.py migrate`).  There should be no errors.
- Make sure that learning resources still have associated departments and offerors, using the new primary key fields.
  ```
  from learning_resources.models import *
  LearningResource.objects.values_list("departments__pk", flat=True).distinct()
  LearningResource.objects.values_list("offered_by__pk", flat=True).distinct()
  ```
- Run a backpopulate mgmt command or two to make sure they still work.
 ```
  ./manage.py backpopulate_ocw_data --overwrite --course-name 6-001-structure-and-interpretation-of-computer-programs-spring-2005
  ./manage.py backpopulate_prolearn_data
```
